src/algebra/monoid_algebra/no_zero_divisors.lean

```diff
@@ -0,0 +1,132 @@
+/-
+Copyright (c) 2022 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import algebra.monoid_algebra.basic
+
+/-!
+# Variations on non-zero divisors in `add_monoid_algebra`s
+
+This file proves that the `add_monoid_algebra R A` has no zero-divisors under the following
+assumptions
+* the semiring `R` itself has no zero-divisors;
+* the grading type `A` has an addition, a linear order and both addition on the left and addition
+  on the right are strictly monotone functions.
+
+The eventual goal is to weaken the assumptions on `A`.  For instance, imposing the same monotonicity
+assumptions on addition, but requiring the order on `A` to be a partial_order instead of a linear
+order should be enough:
+* TODO: a linear extension preserving the monotonicity assumptions should always exist.
```

Comment:
```suggestion
The eventual goal is to weaken the assumptions on `A`.
```
Not so sure about this: equality is a partial order, satisfies `covariant_class X X (+) (<)` due to lack of candidates, but the result does not apply to all grading types.